### PR TITLE
Feature/support assume role with web identity

### PR DESCRIPTION
**Pull request by markdouthwaite**
## This PR includes:

* Addition of `assume_role_with_web_identity` function to fetch credentials from STS. 
* Update to `locate_credentials` to check Web Identity env vars _before_ proceeding to other env vars.
* Bump version to `0.6.1`.

This enables users of the `aws.s3` package to continue to access AWS services given changes to how credentials are fetched/managed in Workflows.

This has been tested on the `newtest` tenant on the `test` environment, and successfully reads, writes and deletes S3 objects on that tenant.

To use this update, users should install `aws.signature` from this repository rather than the existing CRAN/MRAN versions. Alternatively, this PR could be committed back to `cloudyr`, though we don't control that release cycle. If installed this way, _DS should not need to make any further changes to code_. 

## Next steps:

* This PR needs to be reviewed and tested and signed-off on by members of the DS team familiar with R.

### CODE REVIEW

**Review by parul157 — APPROVED** [review]
working as expected 👍

**Review by azhar22k — COMMENTED** [review]
Looks good to me.
keep up the good work

**Review by azhar22k — APPROVED** [review]

File: R/assume_role.R [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+assume_role_with_web_identity <- function(
+  role_arn, 
+  token_file, 
+  base_url=Sys.getenv("AWS_STS_ENDPOINT", "https://sts.amazonaws.com"), 
+  session_name=NULL, 
+  duration=3600,
+  version="2011-06-15"
+){
+  if (is.null(session_name)) {
+    # todo: add random uuid string to the end? (MD)
```
Comment (lukedyer-uipath) [inline R/assume_role.R]: @markdouthwaite: TODO should be enacted or removed

File: R/locate_credentials.R [hunk 2]
```diff
@@ -286,11 +286,27 @@ check_for_user_supplied_profile <- function(profile, file, region, session_token
...
+    identity <- list(arn=Sys.getenv("AWS_ROLE_ARN"),
+                     token_file=Sys.getenv("AWS_WEB_IDENTITY_TOKEN_FILE"))
+    
+    if (!is_blank(identity$arn) && !is_blank(identity$token_file)){
+      creds <- assume_role_with_web_identity(identity$arn, identity$token_file)
+      
+      # this is a tad ugly... (MD)
+      key <- creds$AssumeRoleWithWebIdentityResponse$AssumeRoleWithWebIdentityResult$Credentials$AccessKeyId
+      secret <- creds$AssumeRoleWithWebIdentityResponse$AssumeRoleWithWebIdentityResult$Credentials$SecretAccessKey
+      session_token <- creds$AssumeRoleWithWebIdentityResponse$AssumeRoleWithWebIdentityResult$Credentials$SessionToken
```
Comment (lukedyer-uipath) [inline R/locate_credentials.R]: Should resolve this comment:
Could do it via 
```suggestion
      creds <- creds$AssumeRoleWithWebIdentityResponse$AssumeRoleWithWebIdentityResult$Credentials
      key <- creds$AccessKeyId
      secret <- creds$SecretAccessKey
      session_token <- creds$SessionToken
```

File: R/assume_role.R [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    RoleArn=role_arn,
+    RoleSessionName=session_name,
+    WebIdentityToken=token,
+    Version=version
+  )
+
+  response <- httr::GET(base_url, query=query)
+  content <- httr::content(response)
+
+  if (httr::status_code(response) == 200) {
```
Comment (lukedyer-uipath) [inline R/assume_role.R:39]: - Does it need to be `200` or are we happy with `20x`?
 - Will `httr` follow redirects and do we want it to?

File: R/assume_role.R [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  role_arn, 
+  token_file, 
+  base_url=Sys.getenv("AWS_STS_ENDPOINT", "https://sts.amazonaws.com"), 
+  session_name=NULL, 
+  duration=3600,
+  version="2011-06-15"
+){
+  if (is.null(session_name)) {
+    # todo: add random uuid string to the end? (MD)
+    session_name <- Sys.getenv("TENANT", "DataScienceAssumeRoleWithWebIdentity")
```
Comment (lukedyer-uipath) [inline R/assume_role.R]: This wouldn't work for an upstream contribution to `aws.signature`.  Should it be changed to be compatible with a PR to `aws.signature`?

**Review by lukedyer-uipath — COMMENTED** [review]
Some clean up suggestions

File: R/locate_credentials.R (hunk 2 above)
Comment (lukedyer-uipath) [reply R/locate_credentials.R]: Also will need to delete the code comment too

File: R/assume_role.R (hunk 3 above)
Comment (markdouthwaite) [reply R/assume_role.R:39]: The AWS docs don't list any other successful return status other than 200 in this case.

I don't believe we'll want to follow redirects in this case.

**Comment by markdouthwaite** [comment]
@parul157 - could you give this a quick once-over to check you're happy? 

Peak-specific env var (`TENANT`) used for the `RoleSessionName` param has been dropped in favour of extracting resource ID from the ARN.

I've also tidied up a little duplication and dropped comments, too.

cc @lukedyer-peak

**Review by parul157 — APPROVED** [review]
Looks good, works perfectly. 👍

File: R/assume_role.R [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#' @rdname assume_role_with_web_identity
+#' @title Assume Role with AWS Web Identity
+#' @description Assume a role from a provided Web Identity Token and ARN using AWS Secure Token Service (STS).
+#' @param role_arn A character string containing the AWS Role Amazon Resource Name (ARN). This specifies the permissions you have to access other AWS services.
+#' @param token_file A character string containing a path to a Web Identity Token file.
+#' @param base_url The AWS STS endpoint to use to retrieve your credentials from.
+#' @param session_name A character string optionally specifying the name.
+#' @param duration The expiry time on the retrieved credentials.
```
Comment (lukedyer-uipath) [inline R/assume_role.R:8]: duration is measured in `seconds`? Might be worth adding to docstring

**Review by lukedyer-uipath — APPROVED** [review]
LGTM: small docstring clarification

**Pull request merged**